crypto/ts/quinTree.ts

```diff
@@ -0,0 +1,377 @@
+    fullPath.leaf = subTree.root;
+
+    return fullPath;
+  }
+
+  /**
+   * Verify a proof
+   * @param proof The proof to verify
+   * @returns Wether the proof is valid
+   */
+  verifyProof = (proof: IMerkleProof): boolean => {
+    const { pathElements, leaf, root, pathIndices } = proof;
+
+    // Hash the first level
+    const firstLevel: bigint[] = pathElements[0].map(BigInt);
+
+    firstLevel.splice(Number(pathIndices[0]), 0, leaf);
+
+    let currentLevelHash: bigint = this.hashFunc(firstLevel);
+
+    // Verify the proof
+    for (let i = 1; i < pathElements.length; i += 1) {
+      const level: bigint[] = pathElements[i].map(BigInt);
+      level.splice(Number(pathIndices[i]), 0, currentLevelHash);
+
+      currentLevelHash = this.hashFunc(level);
+    }
+
+    // the path is valid if the root matches the calculated root
+    return currentLevelHash === root;
+  };
+
+  /**
+   * Calculate the indices of the parent
+   * @param index The index of the leaf
+   * @returns The indices of the parent
+   */
+  calcParentIndices(index: number): number[] {
+    // can only calculate the parent for leaves not subroots
+    if (index >= this.capacity) {
+      throw new Error(`Index ${index} is out of bounds`);
+    }
+
+    const indices: number[] = [];
+    let r = index;
+    let levelCapacity = 0;
+    for (let i = 0; i < this.depth; i += 1) {
+      levelCapacity += this.arity ** (this.depth - i);
+      r = Math.floor(r / this.arity);
+      const levelIndex = levelCapacity + r;
+      indices.push(levelIndex);
+    }
+
+    return indices;
+  }
+
+  /**
+   * Calculate the indices of the children of a node
+   * @param index The index of the node
+   * @returns The indices of the children
+   */
+  calChildIndices(index: number): number[] {
+    // cannot get the children of a leaf
+    if (index < this.capacity) {
+      throw new Error(`Index ${index} is out of bounds. Can only get children of subroots`);
+    }
+
+    // find the level
+    let level = 0;
+    let r = this.arity ** level;
+    do {
+      level += 1;
+      r += this.arity ** level;
+    } while (index >= r);
+
+    const start = (index - this.arity ** level) * this.arity;
+    const indices = Array<number>(this.arity)
+      .fill(0)
+      .map((_, i) => start + i);
+
+    return indices;
+  }
+
+  /**
+   * Get a node at a given index
+   * @param index The index of the node
+   * @returns The node
+   */
+  getNode(index: number): Leaf {
+    // if we have it, just return it
+    if (this.nodes[index]) {
+      return this.nodes[index];
+    }
+
+    // find the zero value at that level
+    // first need to find the level
+    let r = 0;
+    let level = this.depth;
+
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition, no-constant-condition
+    while (true) {
+      r += this.arity ** level;
+      if (index < r) {
+        break;
+      }
+
+      level -= 1;
+    }
```

Comment:
wouldn't work, though I managed to remove while(true) for another condition, lmk if that looks ok

Reply:
> wouldn't work, though I managed to remove while(true) for another condition, lmk if that looks ok

It's good, but if you can get rid of break it will be better :)